datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -306,37 +307,99 @@ func (c *Consumer) updateOffset(buffer [][]*Segment, lastMsg *sarama.ConsumerMes
 	return nil
 }
 
+// renewExpireMsg updates expiration timer and message if the expiring message is handled.
+func (c *Consumer) renewExpireMsg(buffer [][]*Segment, timer *time.Timer, oldestMsg *sarama.ConsumerMessage) *sarama.ConsumerMessage {
+	if c.config.ExpirationTime > time.Duration(0) {
+		if len(buffer) == 0 {
+			timer.Stop()
+			return nil
+		} else if oldestMsg != buffer[0][0].orig {
+			timer.Reset(c.config.ExpirationTime)
+			return buffer[0][0].orig
+		} else {
+			return oldestMsg
+		}
+	}
+	return nil
+}
+
 // ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
 // Once the Messages() channel is closed, the Handler must finish its processing
 // loop and exit.
 func (c *Consumer) ConsumeClaim(cgs sarama.ConsumerGroupSession, cgc sarama.ConsumerGroupClaim) error {
-	var buffer [][]*Segment
-	for msg := range cgc.Messages() {
-		if len(buffer) > c.config.MaxMessageNumber {
-			return fmt.Errorf("%v: increasing buffer size may resolve this problem. [max: %v, current: %v]", bufferOverflowErrorMsg, c.config.MaxMessageNumber, len(buffer))
-		}
+	if buffer, err := c.consumeClaim(cgs, cgc); err != nil {
+		return c.handleError(buffer, cgs, err)
+	}
+	return nil
+}
 
-		segment, err := newSegment(msg)
-		if err != nil {
-			return err
-		}
+func (c *Consumer) consumeClaim(cgs sarama.ConsumerGroupSession, cgc sarama.ConsumerGroupClaim) ([][]*Segment, error) {
+	var (
+		buffer      [][]*Segment // TODO-Chaindatafetcher better to introduce segment buffer structure with useful methods
+		expireMsg   *sarama.ConsumerMessage
+		expireTimer = time.NewTimer(c.config.ExpirationTime)
+	)
 
-		// insert a new message segment into the buffer
-		buffer, err = insertSegment(segment, buffer)
-		if err != nil {
-			return err
-		}
+	// make sure that the expireTimer channel is empty
+	if !expireTimer.Stop() {
+		<-expireTimer.C
+	}
 
-		// handle the buffered messages if any message can be reassembled
-		buffer, err = c.handleBufferedMessages(buffer)
-		if err != nil {
-			return err
+	for {
+		select {
+		case <-expireTimer.C:
+			return buffer, errors.New(msgExpiredErrorMsg)
+		case msg, ok := <-cgc.Messages():
+			if !ok {
```

Comment:
@jeongkyun-oh It returns `buffer, nil` when ok is false.. is it okay to return like that?

Reply:
@yoomee1313 here, okay is false when the channel is closed. So, if message channel is closed, then it is returned as a normal case.
https://tour.golang.org/concurrency/4

Reply:
Oh, i didn't know that channel closing is normal case. Thanks!